### ISSUE TITLE
Training progress

### ISSUE THREAD

**Issue by MMayla** [axa-group/nlp.js#206]
**Is your feature request related to a problem? Please describe.**
I train on pretty big data, more than 10k+ questions, It's pretty frustrating to just wait for days till training ends without knowing how much progress the training made, 

**Describe the solution you'd like**
I suggest a way to track training progress so I will be able to print something like this
```
# train question (1/1000)
# train question (2/1000)
# train question (3/1000)
.
.
.
# train question (999/1000)
# train question (1000/1000)
```

**Describe alternatives you've considered**
Let's share ideas of ways to achieve the same thing as a workaround currently or as a feature

**Comment by jesus-seijas-sp** [comment]
Is not so simple when we are talking about neural networks... Is not a for loop when you can know the number of iterations.

This neural network has 3 exit conditions: maximum number of epochs reached, acceptable loss reached, or delta error loss too low compared with the previous epoch. But you cannot know when this will happen. You can estimate it, but taking into account that the training proccess is a core function that right know is fully optimized (passed from 36 minutes to train 500 intents to less than 10 seconds) adding this estimation will make this process to be too much slower. Adding console logs also affects to the performance.

But also an analysis of your business case: take into account that platforms like Microsoft LUIS have a limitation of 500 intents and 2500 utterance, and they are training in servers with GPU or TPU. The Neural Network in NLP.js is optimized for CPU in order to train even faster that other providers with GPU. I train models with more than 500 intents in less than 10 seconds... because I split the models into domains, so are trained separately.

The mathematics behind: Imagine that you have 1000 intents with 5000 utterances (5 per intent), then you have to calculate the features: each different stem, that is, different words but taking into account that derivated words count as the same feature ("developer", "developing", "developed"... are the same feature). Imagine that there are 500 features. That means that the neural network will have 501 *  1000 = 501000 variables (features * intents + intents) that haves to modify in order to train. This is the order of complexity in each epoch for each utterance, so to measure the time per epoch the scale will be in the order of 5000 * 501000 = 2.505.000.000. This is too much.

But it can be splitted into domains: a domain is a set of data of the training that is similar between it because belongs to the same functional domain. Imagine that you have 2 domains, so the numbers are more or less: each domain has 2500 utterance, 500 intents, and 300 features each (some of the features are shared) but 500 unique features.
If we do the same calculations for each domain we will find that the number of variables is 300 * 500 + 500 = 150.500, so the order of time to train each domain is 2500 * 150.500 = 376.250.000.
So multiplied by 2 (2 domains) the scale of time to train is 752.500.00 that is 33 times lower than the previous situation with only one domain. 
But we need also a master domain, that given an utterance returns the domain to use, in this case there will be 2 intents (the domains), 5000 utterances and 500 features, to a total of 1002 variables and the scale of time is 5.010.000. Even with this extra domain, the training is 30 times lower.

Also, theorically talking, as 200 intents per domain have a good timing and score, you can have 200*200= 40.000 different intents trained in 200 different domains with a master domain, working without too much score lose.

You have an example of domains here: https://github.com/axa-group/nlp.js/blob/7186a9016657d62d33e7140b479abbf283000c06/docs/nlu-manager.md

The NlpManager also haves the assignDomain method, so the same applies to NlpManager.

**Comment by MMayla** [comment]
@jesus-seijas-sp I really appreciate taking the time to write this, I have learned a lot from it. You have opened my eyes to ways that can speed up the training time in my case.

Thanks again for the amazing explanation, I suggest to put this explanation in for example "behind the scene" section in the docs, it will help people understand a lot.

I adore the work you did in this project, it is really outstanding.

**Comment by MMayla** [comment]
@jesus-seijas-sp as a follow-up regarding what you have suggested.

Before version 3.1.1, I didn't see the end of the training, I waited +6 hours and every time training crash, so I had to train on a subset of the intents which reduced training time to 3 hours, definitely what I didn't want (I want to train on the whole data), but at least I could kick start the project with that.

After version 3.1.1, I finally got to train the whole data without program crashing, and that wasn't the greatest part yet, it did train with only 1 hour of training,  and yes definitely it's performance, performance, performance :smiley: 

After your great explanation and suggestion using domains, I didn't believe myself training finished in just **11 minutes** :dancer: 

Thank you :blush:
